ably/realtime_conn.go

```diff
@@ -519,86 +527,92 @@ func (c *Connection) eventloop() {
 				c.callbacks.onChannelMsg(msg)
 				break
 			}
+
 			c.state.Lock()
-			if c.reconnecting {
-				c.reconnecting = false
-				if tokenError(msg.Error) {
-					// (RTN15c5)
-					// TODO: (gernest) implement (RTN15h) This can be done as a separate task?
+			reauthorizing := c.reauthorizing
+			c.reauthorizing = false
+			if isTokenError(msg.Error) {
+				if reauthorizing {
+					c.lockedReauthorizationFailed(newErrorProto(msg.Error))
+					c.state.Unlock()
+					return
 				} else {
-					// (RTN15c4)
-					c.callbacks.onReconnectMsg(msg)
+					// TODO: RTN14b; may reuse c.reauthorize from RTN15h2.
 				}
 			}
-			c.setState(StateConnFailed, newErrorProto(msg.Error))
 			c.state.Unlock()
-			c.queue.Fail(newErrorProto(msg.Error))
-			if c.conn != nil {
-				c.conn.Close()
-			}
+
+			c.failedConnSideEffects(msg.Error)
 		case proto.ActionConnected:
 			c.auth.updateClientID(msg.ConnectionDetails.ClientID)
 			c.state.Lock()
 			// we need to get this before we set c.key so as to be sure if we were
 			// resuming or recovering the connection.
 			mode := c.getMode()
-			c.state.Unlock()
 			if msg.ConnectionDetails != nil {
-				c.state.Lock()
 				c.key = msg.ConnectionDetails.ConnectionKey //(RTN15e) (RTN16d)
-				c.state.Unlock()
 
 				// Spec RSA7b3, RSA7b4, RSA12a
 				c.auth.updateClientID(msg.ConnectionDetails.ClientID)
 
 				maxIdleInterval := time.Duration(msg.ConnectionDetails.MaxIdleInterval) * time.Millisecond
 				receiveTimeout = c.opts.realtimeRequestTimeout() + maxIdleInterval // RTN23a
 			}
-			c.state.Lock()
 			reconnecting := c.reconnecting
 			if reconnecting {
 				// reset the mode
 				c.reconnecting = false
+				c.reauthorizing = false
 			}
-			id := c.id
+			previousID := c.id
+			c.id = msg.ConnectionID
+			c.msgSerial = 0
+			if reconnecting && mode == recoveryMode {
+				// we are setting msgSerial as per (RTN16f)
+				msgSerial, err := strconv.ParseInt(strings.Split(c.opts.Recover, ":")[2], 10, 64)
+				if err != nil {
+					//TODO: how to handle this? Panic?
```

Comment:
We will need to figure out our logging situation at some point. I think implementing the spec is more important at the moment, so its fine thisTODO stays.

Reply:
I just moved this code around, the TODO was already there.

Anyway, I agree we should improve logging (see #160) once the codebase becomes less of a moving target.

In this particular case, though, I think a panic is in order. A malformed msgSerial just shouldn't ever happen, and it's already panicking if it doesn't have at least two parts separated by `:`.

Reply:
I see, if that is a case you can just as well add a panic here as part of this PR. We don't need to keep that TODO anymore.